lib/Dialect/FIRRTL/Transforms/RegisterOptimizer.cpp

```diff
@@ -0,0 +1,149 @@
+//===- RegisterOptimizer.cpp - Register Optimizer ---------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// This pass optimized registers as allowed by historic firrtl register
+// behaviors.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "mlir/IR/Dominance.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "firrtl-register-optimizer"
+
+using namespace circt;
+using namespace firrtl;
+
+// Instantiated for RegOp and RegResetOp
+template <typename T>
+static bool canErase(T op) {
+  return !(hasDontTouch(op.getResult()) ||
+           (op.getAnnotationsAttr() && !op.getAnnotationsAttr().empty()));
+}
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+struct RegisterOptimizerPass
+    : public RegisterOptimizerBase<RegisterOptimizerPass> {
+  void runOnOperation() override;
+  void checkRegReset(mlir::DominanceInfo &dom,
+                     SmallVector<Operation *> &toErase, RegResetOp reg);
+  void checkReg(mlir::DominanceInfo &dom, SmallVector<Operation *> &toErase,
+                RegOp reg);
+};
+
+} // namespace
+
+void RegisterOptimizerPass::checkReg(mlir::DominanceInfo &dom,
+                                     SmallVector<Operation *> &toErase,
+                                     RegOp reg) {
+  if (!canErase(reg))
+    return;
+  auto con = getSingleConnectUserOf(reg.getResult());
+  if (!con)
+    return;
+
+  // Register is only written by itself, replace with invalid.
+  if (con.getSrc() == reg) {
+    auto inv =
+        OpBuilder(reg).create<InvalidValueOp>(reg.getLoc(), reg.getType());
+    reg.replaceAllUsesWith(inv.getResult());
+    toErase.push_back(reg);
+    toErase.push_back(con);
+    return;
+  }
+  // Register is only written by a constant
+  if (isConstant(con.getSrc())) {
+    // constant may not dominate the register.  But it might be the next
+    // operation, so we can't just move it.  Straight constants can be
+    // rematerialized.  Derived constants are piped through wires.
+    bool dominatesAll = true;
+    for (auto *use : reg->getUsers()) {
+      if (use == con)
+        continue;
+      if (!dom.dominates(con.getSrc(), use)) {
```

Comment:
`dom.dominates` calls `isBeforeInBlock` internally so it would  be quadratic time in the worst case. 
https://github.com/llvm/llvm-project/blob/c0967995d254fd2dd24ba74afa46df5f559f11ba/mlir/lib/IR/Dominance.cpp#L285 

Not ideal but I think it's better to always create constants at the beginning of the block.

Reply:
It should be on average O(1) for ssa regions per the isBeforeInBlock comments.  I believe it's a DFS numbering internally (which is stored on the operation).

Reply:
`cst->moveBefore` invalidates numberings (at least when I checked 4 month ago) and it requires O(N) recomputation every time.